crates/sui-core/src/authority_server.rs

```diff
@@ -622,6 +616,160 @@ impl Validator for ValidatorService {
+        });
+    }
+}
+
+/// Implements generic pre- and post-processing. Since this is on the critical
+/// path, any heavy lifting should be done in a separate non-blocking task
+/// unless it is necessary to override the return value.
+#[macro_export]
+macro_rules! handle_with_decoration {
+    ($self:ident, $func_name:ident, $request:ident) => {{
+        // extract IP info. Note that in addition to extracting the client IP from
+        // the request header, we also get the remote address in case we need to
+        // throttle a fullnode, or an end user is running a local quorum driver.
+        let connection_ip: Option<SocketAddr> = $request.remote_addr();
+
+        // This should never happen except perhaps in simtest or some other non-standard
+        // environment. If we are seeing this, we should investigate
+        // TODO: add metric here
+        if connection_ip.is_none() {
+            if cfg!(all(test, not(msim))) {
+                panic!("Failed to get remote address from request");
+            } else {
+                error!("Failed to get remote address from request");
+            }
+        }
+
+        let proxy_ip: Option<SocketAddr> =
+            if let Some(op) = $request.metadata().get("x-forwarded-for") {
+                match op.to_str() {
+                    Ok(ip) => match ip.parse() {
+                        Ok(ret) => Some(ret),
+                        Err(e) => {
+                            error!("Failed to parse x-forwarded-for header value to SocketAddr: {:?}", e);
+                            None
+                        }
+                    },
+                    Err(e) => {
+                        error!("Invalid UTF-8 in x-forwarded-for header: {:?}", e);
+                        None
+                    }
+                }
+            } else {
+                None
+            };
+
+        // check if either IP is blocked, in which case return early
+        $self.handle_traffic_req(connection_ip, proxy_ip).await?;
+        // handle request
+        let response = $self.$func_name($request).await;
+        // handle response tallying
+        $self.handle_traffic_resp(connection_ip, proxy_ip, &response);
+        response
+    }};
+}
+
+#[async_trait]
+impl Validator for ValidatorService {
+    async fn transaction(
+        &self,
+        request: tonic::Request<Transaction>,
+    ) -> Result<tonic::Response<HandleTransactionResponse>, tonic::Status> {
+        let validator_service = self.clone();
+
+        // Spawns a task which handles the transaction. The task will unconditionally continue
+        // processing in the event that the client connection is dropped.
+        spawn_monitored_task!(async move {
+            // let service_clone = validator_service.clone();
+            // NB: traffic tally wrapping handled within the task rather than on task exit
+            // to prevent an attacker from subverting traffic control by severing the connection
```

Comment:
this is an excellent point, but doesn't it apply to all the methods below as well?